src/daggy.js

```diff
@@ -1,127 +1,136 @@
-    const fields = [].slice.apply(arguments);
-
-    function toString(args) {
-      const x = [].slice.apply(args);
-      return () => {
-        const values = x.map((y) => y.toString());
-        return '(' + values.join(', ') + ')';
-      };
+const { toString } = require('sanctuary-type-classes')
+const type = require('sanctuary-type-identifiers')
+
+// Name of prop used to store:
+// * name of variant of a sum type
+const TAG = '@@tag'
+// * array of arguments used to create a value (to speed up `cata`)
+const VALUES = '@@values'
+
+// adopted version of withValue from  https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/Object/defineProperty
+const defProp = (obj, prop, val) => {
+  var desc = defProp.desc || (
+    defProp.desc = {
+      enumerable: false,
+      writable: false,
+      configurable: false,
+      value: null
     }
+  )
+  desc.value = val
+  Object.defineProperty(obj, prop, desc)
+}
 
-    function wrapped() {
-        const self = getInstance(this, wrapped);
-        var i;
-
-        if(arguments.length != fields.length)
-            throw new TypeError('Expected ' + fields.length + ' arguments, got ' + arguments.length);
-
-        for(i = 0; i < fields.length; i++)
-            self[fields[i]] = arguments[i];
+// optimised version of `arr.map(toString).join(', ')`
+const arrToString = (arr) => {
+  if (arr.length === 0) {
+    return ''
+  }
+  let str = '(' + toString(arr[0])
+  for (var i = 1; i < arr.length; i++) {
+    str += ', ' + toString(arr[i])
+  }
+  return str + ')'
+}
 
-        self.toString = toString(arguments);
+const tagged = (typeName, fields) => {
+  const proto = {}
+  proto.toString = tagged$objToString
+  // this way we avoid named function
+  const typeRep = (0, (...args) => makeValue(fields, proto, args))
+  typeRep.toString = typeRepToString
+  typeRep.prototype = proto
+  typeRep.is = isType
+  typeRep['@@type'] = typeName
+  proto.constructor = typeRep
+  return typeRep
+}
 
-        return self;
+// [1] - possible other names are: `hasInstance`, `isVariant`
+// [2] - it might not be the best choise to also define is on value as
+//       user could use `===` as there should only be one such value.
+const taggedSum = (typeName, definitions) => {
+  const proto = {}
+  proto.cata = taggedSum$cata
+  proto.toString = taggedSum$objToString
+  const typeRep = {
+    toString: typeRepToString,
+    prototype: proto,
+    is: isType, // [1]
+    '@@type': typeName
+  }
+  proto.constructor = typeRep
+  Object.keys(definitions).forEach(tag => {
+    const fields = definitions[tag]
+    const tagProto = Object.create(typeRep.prototype)
+    defProp(tagProto, TAG, tag)
+    if (fields.length === 0) {
+      typeRep[tag] = makeValue(fields, tagProto, [])
+      typeRep[tag].is = taggedSum$isUnitValue // [1,2]
+      return
     }
-    wrapped._length = fields.length;
-    return wrapped;
+    typeRep[tag] = (...args) => makeValue(fields, tagProto, args)
```

Comment:
This might be optimized by pre-compiling a `makeValue` functions with the correct arity up to a certain arity.